### ISSUE TITLE
Assert error in sml_trimstore()

### ISSUE THREAD

**Issue by acachy** [varnishcache/varnish-cache#3816]
```
varnishd  Child (12005) Panic at: Thu, 16 Jun 2022 00:57:57 GMT
Assert error in sml_trimstore(), storage/storage_simple.c line 539:
  Condition((oc->boc->stevedore_priv) == 0) not true.
version = varnish-7.0.2 revision 9b5f68e19ca0ab60010641e305fd12822f18d42c, vrt api = 14.0
ident = Linux,5.15.8,x86_64,-junix,-sfile,-sfile,-smalloc,-smalloc,-sdefault,-hcritbit,epoll
now = 12179286.306722 (mono), 1655341076.980677 (real)
Backtrace:
  0x55d7285a47d5: /usr/sbin/varnishd(+0x637d5) [0x55d7285a47d5]
  0x55d7285a4ee7: /usr/sbin/varnishd(+0x63ee7) [0x55d7285a4ee7]
  0x55d72864b84f: /usr/sbin/varnishd(VAS_Fail+0x59) [0x55d72864b84f]
  0x55d72861c645: /usr/sbin/varnishd(+0xdb645) [0x55d72861c645]
  0x55d7285a0023: /usr/sbin/varnishd(ObjExtend+0x2ca) [0x55d7285a0023]
  0x55d72858e1ab: /usr/sbin/varnishd(VFP_Extend+0xa8) [0x55d72858e1ab]
  0x55d72858baeb: /usr/sbin/varnishd(+0x4aaeb) [0x55d72858baeb]
  0x55d72861b29c: /usr/sbin/varnishd(+0xda29c) [0x55d72861b29c]
  0x55d72859fb46: /usr/sbin/varnishd(ObjIterate+0x130) [0x55d72859fb46]
  0x55d72858c0ba: /usr/sbin/varnishd(+0x4b0ba) [0x55d72858c0ba]
```

``
-p tcp_fastopen=on -p vcl_cooldown=1 -p vsm_free_cooldown=10 -p rush_exponent=2 -p connect_timeout=3.5 -p vsl_mask=-ExpKill,-VCL_trace,-WorkThread,-Hash,-VfpAcct,-ReqProtocol,-RespProtocol,-RespReason,-VCL_use -p workspace_backend=128k -p cli_limit=256k -p cli_timeout=9900 -p ping_interval=0 -p max_retries=7 -p thread_pool_destroy_delay=2 -p thread_pool_reserve=0 -p thread_pool_timeout=600 -p thread_queue_limit=1000 -p nuke_limit=150 -p lru_interval=2 -p send_timeout=14400 -p idle_send_timeout=75 -p ban_lurker_sleep=0.2 -p ban_lurker_batch=15 -p ban_lurker_holdoff=2 -p ban_lurker_age=1 -p thread_stats_rate=20 -p gzip_level=9 -p listen_depth=16383 -p clock_step=1500 -p fetch_chunksize=16k -p gzip_memlevel=8 -p gzip_buffer=32k -p vsl_buffer=4k -p http_max_hdr=64 -p http_req_hdr_len=16k -p http_resp_hdr_len=16k -p workspace_client=64k -p timeout_idle=4 -p backend_idle_timeout=60 -p thread_pool_min=162 -p thread_pool_max=1500 -p thread_pools=2 -s disk0=file,/d0/d.bin,33750m -s disk1=file,/d1/d.bin,33750m -s mems=malloc,500m -s junk=malloc,30m
``

- Multiple servers with significant load, huge amount of simultaneous ban() requests, many backends, 1mb+ vcl, etc.
- Normally objects are distributed between several file-storages (by random) and everything works OK
- If I add conditions to send some small objects to malloc storage ("set beresp.storage=storage.mems") the panic start happening each 5-10 minutes (probably even before any storage become fully filled)
- Additional information is problematic for now and at least will take a lot of time

**Comment by acachy** [comment]
Sorry, my logging system doesn't keep full logs by default, so that's just all information I have now. And to repeate problem in non-production environment at least would take some development to mirror requests.

That's why - I'd be able to tell extra details only after some time.

**Comment by qdupuy** [comment]
Same issue :/

```
Debug: Version: varnish-7.0.2 revision 9b5f68e19ca0ab60010641e305fd12822f18d42c
Debug: Platform: Linux,5.4.0-91-generic,x86_64,-jnone,-smalloc,-sfile,-sdefault,-hcritbit
Debug: Child (114) Started
Info: Child (114) said Child starts
Info: Child (114) said SMF.static mmap'ed 2147483648 bytes of 2147483648
Error: Child (114) died signal=6 (core dumped)
Error: Child (114) Panic at: Fri, 09 Sep 2022 12:40:17 GMT
Assert error in sml_trimstore(), storage/storage_simple.c line 539:
  Condition((oc->boc->stevedore_priv) == 0) not true.
version = varnish-7.0.2 revision 9b5f68e19ca0ab60010641e305fd12822f18d42c, vrt api = 14.0
ident = Linux,5.4.0-91-generic,x86_64,-jnone,-smalloc,-sfile,-sdefault,-hcritbit,epoll
now = 22976411.936069 (mono), 1662727217.145549 (real)
Backtrace:
  0x564a78b8282c: varnishd(+0x5882c) [0x564a78b8282c]
  0x564a78bfc298: varnishd(VAS_Fail+0x18) [0x564a78bfc298]
  0x564a78bd9fad: varnishd(+0xaffad) [0x564a78bd9fad]
  0x564a78b6dff6: varnishd(+0x43ff6) [0x564a78b6dff6]
  0x564a78bda469: varnishd(+0xb0469) [0x564a78bda469]
  0x564a78b6db35: varnishd(+0x43b35) [0x564a78b6db35]
  0x564a78b6e7b6: varnishd(+0x447b6) [0x564a78b6e7b6]
  0x564a78ba7d77: varnishd(+0x7dd77) [0x564a78ba7d77]
  0x564a78ba82d4: varnishd(+0x7e2d4) [0x564a78ba82d4]
  0x7fa0a7bd3ea7: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8ea7) [0x7fa0a7bd3ea7]
argv = {
  [0] = \"varnishd\",
  [1] = \"-F\",
  [2] = \"-f\",
  [3] = \"/etc/varnish/default.vcl\",
  [4] = \"-a\",
  [5] = \"proxy=:8443,PROXY\",
  [6] = \"-p\",
  [7] = \"feature=+http2\",
  [8] = \"-s\",
  [9] = \"malloc,512M\",
  [10] = \"-a\",
  [11] = \":8080\",
  [12] = \"-s\",
  [13] = \"static=file,/varnish-cache,2G\",
}
```

**Comment by acachy** [comment]
Finally I was able to catch full log
It's already a bit different settings and newer version (master #f1d450c) 
[varn-panic.txt](https://github.com/varnishcache/varnish-cache/files/9984934/varn-panic.txt)

**Comment by hont512** [comment]
A little addition
WIth another similar type of size-related logic, using different storage types, I've got another stable error:
```
varnishd  Child (777801) Panic at: Fri, 28 Nov 2022 10:24:42 GMT#012Wrong turn at cache/cache_main.c:325:#012Signal 11 (Segmentation fault) received at (nil) si_code 1#012version = varnish-trunk revision f1d450cfa57862e080c1f54ec6a3a521bbd192be, vrt api = 16.0
```
Not sure if related, but maybe some hint. I'll post as separate issue if still persist after this ticket's solution.

**Comment by Dridi** [comment]
Could this be a consequence of https://github.com/varnishcache/varnish-cache/pull/3557 ?

**Comment by delthas** [comment]
Hi, I just got the same panic in production, some time after upgrading to 7.3.0. [panic.show](https://github.com/varnishcache/varnish-cache/files/13647721/panic-show.txt)

It panicked as it was storing a beresp to a file stevedore. No specific gzip options configured on Varnish or in the VCL; Varnish just sending Accept-Encoding: gzip to the backend; but in contrary to the panic found by @acachy, where the server returned a file encoded with gzip (can be guessed from the xxx and the GZIPBITS in their panic), in my case the server did not even reply with a Content-Encoding: gzip. So this looks unrelated to vgz?

**Comment by delthas** [comment]
I investigated further on my server and saw that Varnish is acting strange before the panic. Could this help with investigating the bug?

From 20:00 to 20:50 Varnish acts strange (see below), then at 20:50 is the panic. Varnish restarts afterwards:

n_expired stops increasing from 20:00, 50 minutes before the panic (even though the kind and amount of traffic stays approximately constant). (Green curves show rate of expired). exp_mailed and exp_received (not shown) continue to increase, though.
![image-20231215-140328](https://github.com/varnishcache/varnish-cache/assets/1863865/552da80e-799a-4e10-8b6e-9acc200e80ed)
*(Second graph is orange=sess_herd blue=sess_conn red=sess_queued yellow=sess_dropped)*

CPU time spent in kernel increases from 20:00, 50 minutes before the panic.
![image-20231215-140618](https://github.com/varnishcache/varnish-cache/assets/1863865/39ebd147-835e-49ad-b2b7-34735c61d006)

I've had the same pattern happen several times: higher CPU, sudden stop of expired, then after a few dozen minutes, sml_trimstore panic.

**Comment by nigoroll** [comment]
Please check if this happens with #4013 in place. "Commit 1" sounds a lot like it fixed this issue.

**Comment by nigoroll** [comment]
We believe this has been fixed with #4013 and there were no reports to the contrary. Please reopen if this issue still surfaces.
